src/isolation/landlock.rs

```diff
@@ -0,0 +1,356 @@
+			FileSandboxMode::Normal => CompatLevel::BestEffort,
+			FileSandboxMode::Strict => CompatLevel::HardRequirement,
+		}
+	}
+
+	/// Internal Landlock enforcement function used by [`Self::apply_landlock_restrictions`].
+	///
+	/// This function directly interfaces with the [`landlock`] crate, and is responsible for
+	/// enforcing the policies defined in [`UhyveVm::new`](crate::vm::UhyveVm::new).
+	fn enforce_landlock(&self) -> Result<RestrictionStatus, RestrictError> {
+		// Target: Ubuntu 24.04 LTS (Linux Kernel: 6.8.0)
+		let abi = ABI::V4;
+		// Used for explicitly whitelisted files (read & write).
+		let access_all: landlock::BitFlags<AccessFs, u64> = AccessFs::from_all(abi);
+
+		// LANDLOCK_ACCESS_FS_REMOVE_FILE and LANDLOCK_ACCESS_FS_REMOVE_DIR do not apply to
+		// whitelisted files themselves, but apply to the contents of directories transitively.
+		// This means that we have to give Uhyve read-only access to the parent directory with
+		// some additional permissions so as to support fs::remove_file and removing the temporary
+		// directory.
+		let access_dir_with_rm: landlock::BitFlags<AccessFs, u64> =
+			AccessFs::ReadDir | AccessFs::RemoveDir | AccessFs::RemoveFile;
+
+		let res_status = Ruleset::default()
+			.handle_access(access_all)?
+			.create()?
+			.add_rules(
+				self.rw_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-write path ruleset for {:#?}", *p);
+						Self::determine_ruleset(PathFd::new(p)?, abi)
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_paths
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only path ruleset for {:#?}", *p);
+						Self::determine_ruleset(PathFd::new(p)?, abi)
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.add_rules(
+				self.ro_dirs
+					.as_slice()
+					.iter()
+					.map::<Result<_, RestrictError>, _>(|p| {
+						debug!("Adding read-only directory ruleset for {:#?}", *p);
+						Ok(PathBeneath::new(PathFd::new(p)?, access_dir_with_rm))
+					}),
+			)?
+			.set_compatibility(self.compat_level)
+			.set_no_new_privs(true)
+			.restrict_self()?;
+
+		debug!(
+			"Landlock ruleset status: {:#?} (no_new_privs: {:#?})",
+			res_status.ruleset, res_status.no_new_privs
+		);
+
+		// Adapted from Mickaël Salaün's Rust Landlock library, which is distributed
+		// under the terms of the MIT and Apache 2.0 licenses.
+		//
+		// For the original source code authored by Mickaël Salaün, see:
+		// https://github.com/landlock-lsm/rust-landlock/blob/07f2a9a7/examples/sandboxer.rs#L133C1-L135C6
+		if res_status.ruleset == RulesetStatus::NotEnforced {
+			panic!(
+				"Landlock is not supported by the running kernel. You can disable Landlock using `--file-isolation none`."
+			);
+		}
+
+		let mut is_already_enabled = LANDLOCK_ENABLED.lock().unwrap();
+		*is_already_enabled = true;
```

Comment:
```suggestion
		*LANDLOCK_ENABLED.lock().unwrap() = true;
```

Also, I think it is nicer if this logic is moved to `apply_landlock_restrictions`, as the check is also present in that fn

Reply:
The idea was to "abort as early as possible if the execution won't be possible later. But it's a micro-optimization, so sure.